decentralized-message-queue/src/DMQ/Protocol/LocalMsgNotification/Type.hs

```diff
@@ -0,0 +1,181 @@
+    -- making a blocking call for more messages.
+    --
+    MsgServerDone
+      :: Message (LocalMsgNotification msg) (StBusy StBlocking) StDone
+
+    MsgClientDone
+      :: Message (LocalMsgNotification msg) StIdle StDone
+
+  type StateAgency  (StBusy blocking)  = ServerAgency
+  type StateAgency  StIdle  = ClientAgency
+  type StateAgency  StDone  = NobodyAgency
+
+  type StateToken = SingMsgNotification
+
+instance ( NFData msg
+         ) => NFData (Message (LocalMsgNotification msg) from to) where
+  rnf (MsgRequest block)  = rnf block
+  rnf (MsgReply a !_more) = rnf a
+  rnf MsgServerDone{}     = ()
+  rnf MsgClientDone{}     = ()
+
+deriving instance (Eq msg) =>
+                  Eq (Message (LocalMsgNotification msg) from to)
+
+deriving instance (Show msg) =>
+                  Show (Message (LocalMsgNotification msg) from to)
+
+
+-- | The kind of the local message notification protocol, and the types of
+-- the states in the protocol state machine.
+--
+-- It is parameterised over the type of messages
+--
+type LocalMsgNotification :: Type -> Type
+type data LocalMsgNotification msg where
+  -- TODO haddock
+  -- |
+  StIdle :: LocalMsgNotification msg
+
+  -- |
+  StBusy :: StBlockingStyle -> LocalMsgNotification msg
+
+  -- | Nobody has agency. The terminal state.
+  --
+  StDone :: LocalMsgNotification msg
+
+instance ( ShowProxy msg
+         ) => ShowProxy (LocalMsgNotification msg) where
+    showProxy _ =
+      "LocalMsgNotification " ++ showProxy (Proxy :: Proxy msg)
+
+
+-- | The kind of blocking state
+--
+type data StBlockingStyle where
+
+  -- | In this sub-state the reply need not be prompt. There is no timeout.
+  --
+  StBlocking    :: StBlockingStyle
+
+  -- | In this state the peer must reply. There is a timeout.
+  --
+  StNonBlocking :: StBlockingStyle
+
+
+-- | The value level equivalent of 'BlockingStyle'.
+--
+-- This is also used in 'MsgRequest' where it is interpreted (and can be
+-- encoded) as a 'Bool' with 'True' for blocking, and 'False' for non-blocking.
+--
+data SingBlockingStyle (k :: StBlockingStyle) where
+  SingBlocking    :: SingBlockingStyle StBlocking
+  SingNonBlocking :: SingBlockingStyle StNonBlocking
+
+deriving instance Eq   (SingBlockingStyle b)
+deriving instance Show (SingBlockingStyle b)
+
+instance NFData (SingBlockingStyle b) where
+  rnf SingBlocking    = ()
+  rnf SingNonBlocking = ()
+
+
+-- | a singleton witness for protocol state
+--
+type SingMsgNotification :: LocalMsgNotification msg
+                         -> Type
+data SingMsgNotification k where
+    SingIdle  :: SingMsgNotification  StIdle
+    SingBusy  :: SingBlockingStyle blocking
+              -> SingMsgNotification (StBusy blocking)
+    SingDone  :: SingMsgNotification  StDone
+
+deriving instance Show (SingMsgNotification k)
+
+
+instance StateTokenI StIdle where stateToken = SingIdle
+instance {-# OVERLAPPABLE #-} StateTokenI (StBusy blocking) where stateToken = error "StBusy blocking not implemented"
```

Comment:
This should be removed
```suggestion
```

Reply:
The solution (if this is an issue at all :grin:) is different, and it's quite possible that's what I did in the branch I mentioned in the TODO.  The instance can be provided by the `MsgRequestTxIds`.
```hs
    MsgRequestTxIds
      :: forall (blocking :: StBlockingStyle) txid tx.
         SingI blocking
      => SingBlockingStyle blocking
      -> NumTxIdsToAck -- ^ Acknowledge this number of outstanding txids
      -> NumTxIdsToReq -- ^ Request up to this number of txids.
      -> Message (TxSubmission2 txid tx) StIdle (StTxIds blocking)
```